augur/filter/_run.py

```diff
@@ -381,23 +381,27 @@ def run(args):
             dropped_samps = list(sequence_strains - valid_strains)
             write_vcf(args.sequences, args.output, dropped_samps)
     elif args.sequences:
-        sequences = read_sequences(args.sequences)
-
         # If the user requested sequence output, stream to disk all sequences
         # that passed all filters to avoid reading sequences into memory first.
         # Even if we aren't emitting sequences, we track the observed strain
         # names in the sequence file as part of the single pass to allow
         # comparison with the provided sequence index.
+        output_handle = None
+
         if args.output:
-            observed_sequence_strains = set()
-            with open_file(args.output, "wt") as output_handle:
-                for sequence in sequences:
-                    observed_sequence_strains.add(sequence.id)
+            output_sequences_context_manager = open_file(args.output, "wt")
+            output_handle = output_sequences_context_manager.__enter__()
 
-                    if sequence.id in valid_strains:
-                        write_sequences(sequence, output_handle, 'fasta')
-        else:
-            observed_sequence_strains = {sequence.id for sequence in sequences}
+        observed_sequence_strains = set()
+        for sequence in read_sequences(args.sequences):
+            observed_sequence_strains.add(sequence.id)
+
+            if args.output:
+                if sequence.id in valid_strains:
+                    write_sequences(sequence, output_handle, 'fasta')
+
+        if args.output:
+            output_sequences_context_manager.__exit__(None, None, None)
```

Comment:
_non-blocking comment_

The use of `__enter__` and `__exit__` stood out to me here, so I was curious if it's possible to have an optional context manager. Based on this [StackOverflow answer](https://stackoverflow.com/a/12168252), seems like it would be possible to do something like 

```python
from contextlib import nullcontext

observed_sequence_strains = set()
output_context_manager = open_file(args.output, "wt") if args.output is not None else nullcontext()
with output_context_manager as output_handle:
    for sequence in read_sequences(args.sequences):
        observed_sequence_strains.add(sequence.id)

        if output_handle is not None:
            if sequence.id in valid_strains:
                write_sequences(sequence, output_handle, 'fasta')
```

Reply:
Great suggestion, that is much cleaner! Updated:

https://github.com/nextstrain/augur/blob/9826c6dcd6830c2323e3615c465d2d9816b70468/augur/filter/_run.py#L390-L397